bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -2338,33 +2338,37 @@ private static boolean checkIsLikeArrayType(Object sourceValue, BArrayType targe
         if (source.getType().getTag() == TypeTags.ARRAY_TAG) {
             Type sourceElementType = ((BArrayType) source.getType()).getElementType();
             if (isValueType(sourceElementType)) {
-                boolean isType = checkIsType(sourceElementType, targetTypeElementType, new ArrayList<>());
 
-                if (isType || !allowNumericConversion || !isNumericType(sourceElementType)) {
-                    return isType;
-                }
-
-                if (isNumericType(targetTypeElementType)) {
+                if (checkIsType(sourceElementType, targetTypeElementType, new ArrayList<>())) {
                     return true;
                 }
 
-                if (targetTypeElementType.getTag() != TypeTags.UNION_TAG) {
-                    return false;
-                }
+                if (allowNumericConversion && isNumericType(sourceElementType)) {
+                    if (isNumericType(targetTypeElementType)) {
+                        return true;
+                    }
 
-                List<Type> targetNumericTypes = new ArrayList<>();
-                for (Type memType : ((BUnionType) targetTypeElementType).getMemberTypes()) {
-                    if (isNumericType(memType) && !targetNumericTypes.contains(memType)) {
-                        targetNumericTypes.add(memType);
+                    if (targetTypeElementType.getTag() != TypeTags.UNION_TAG) {
+                        return false;
                     }
+
+                    List<Type> targetNumericTypes = new ArrayList<>();
+                    for (Type memType : ((BUnionType) targetTypeElementType).getMemberTypes()) {
+                        if (isNumericType(memType) && !targetNumericTypes.contains(memType)) {
```

Comment:
I think we need to improve these checks a bit more - please see https://github.com/ballerina-platform/ballerina-lang/issues/30301. We can fix it separately though.